🔀 Merge dev to prod to introduce new feature

DrHooBs opened the pull request:
This is needed to introduce the newest feature streamMessage to the production instance.

InimicalPart committed:
Merge pull request #17 from Incoverse/prod

DrHooBs committed:
✨ Add StreamMessage Command

Add a stream message command to facilitate the sending of live announcements for the streamer

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+                )
+        );
+        this._loaded = true;
+        return true;
```
A better solution here, which I myself haven't adopted too well but I think its better to start, is to do the following instead of the ``this._loaded = true`` and the ``return true``:

```suggestion
        return super.setup();
```
`super#setup` basically does ``this._loaded = true`` and ``return true``, so it makes it a bit better to read

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+                            name: "stream", value: "stream"
+                        }, {
+                            name:"cancellation", value: "cancellation"})
+                )
```
can we fix the formatting here so it looks a bit better?

```suggestion
                .addStringOption((option) =>
                    option
                        .setName("template")
                        .setDescription("Select a template")
                        .addChoices(
                            {name: "stream", value: "stream"},
                            {name: "cancellation", value: "cancellation"}
                        )
                )
```

InimicalPart replied:
Please also look at other comment regarding templates.

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+
+        //Get Channel List
+        const channels = await interaction.client.guilds.fetch(global.app.config.mainServer).then(guild => guild.channels.fetch())
+        // check if there is a channel that includes "birthdays" in it's name
```
Invalid comment

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+        if (!channels.some((channel) => /live-(streams|notifications|notifs|announcements)/i.test(channel.name) && channel.type == Discord.ChannelType.GuildText)) {
+
+            if (!channels.some((channel) => /announcements/i.test(channel.name) && channel.type == Discord.ChannelType.GuildText)) {
+                interaction.editReply({content: "No channel found with the name `live-streams`, `live-notifications`, `live-notifs`, `live-announcements`, or `announcements`"});
```
I think we could make this a bit more user-friendly but also a bit more professional and PoV-like (I'm not sure if my suggestion will work as expected but if not, you can tweak it so it works)
```suggestion
                interaction.editReply({content: "I was unable to find a valid channel for the stream announcement.\n#- Searched for: `live-streams`, `live-notifications`, `live-notifs`, `live-announcements`, and `announcements`"});
```

DrHooBs replied:
I understand this.... I think I forget that its a chatbot interacting with users not a CLI application

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+        ).getString("message");
+        const time = (
+            interaction.options as Discord.CommandInteractionOptionResolver
+        ).getString("time");
```
Tbh, I don't really like the formatting here, could we fix that? 🙏

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+            !moment(time, "HH:mm").isValid()
+        ) {
+            return await interaction.editReply({
+                content: "Invalid time format. Please use `hh:mm am/pm` or `HH:mm`",
```
This could be a little ambiguous, what if we do it like this?

 ```suggestion
                 content: "Invalid time format. Please use `hh:mm am/pm` or `HH:mm`. Example: `10:34 am`, `23:12`",
```

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+
+        if (template === "cancellation") {
+            await interaction.editReply({
+                content: `Ok sent to ${channel}`,
```
I think an even better message could be:
```suggestion
                content: `Your message has been sent to {channel}` // Add a preview of the message if possible?
```

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+                content: `${role} We're cancelling the stream! ${message} `,
+            });
+            return;
+        }
```
I was looking a bit to having dynamic templates, for example, in the config file you could specify:
```jsonc
{
   ...
   "streamMessages": {
       "cancelled": "[updates-ping], the stream [next-stream-relative] is cancelled due to [reason]. See you all next time!", // @Live Updates, the stream in 2 days is cancelled due to me being sick. See you all next time!"
       "streaming": "[notifs-ping], we are currently live and playing [game]! Come join us over on Twitch! https://twitch.tv/drvem", // @Live Notifications, we are currently live and playing Valheim! Come join us over on Twitch! https://twitch.tv/drvem
    }
   ...
}
```

DrBot would then populate the things on it's own. I know this bot is only gonna be used for DrVem's server so it's like "does the config really matter in this case?", but I think it does, some people might also use these code bits for their own bots so its kinda nice to add some modularity to make it easier, it's also nice to keep all customizable things in the config file, imo

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+            });
+        }
+
+        let isTwelveHour = moment(time, "hh:mm a").isValid();
```
Since this variable is only being used once, what if we do the type checking in the `convertToTimestamp()` function? It will spare some reader's eyes from understanding all parameters

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+            });
+        }
+
+        if (template === "stream") {
+            await interaction.editReply({
+                content: `Ok sent to ${channel}`,
+            });
+
+            await channel.send({
+                content: `${role} We're going live! ${message} at <t:${timestamp}>`,
+            });
+        }
```
I was looking a bit to having dynamic templates, for example, in the config file you could specify:
```jsonc
{
   ...
   "streamMessages": {
       "cancelled": "[updates-ping], the stream [next-stream-relative] is cancelled due to [reason]. See you all next time!", // @Live Updates, the stream in 2 days is cancelled due to me being sick. See you all next time!"
       "streaming": "[notifs-ping], we are currently live and playing [game]! Come join us over on Twitch! https://twitch.tv/drvem", // @Live Notifications, we are currently live and playing Valheim! Come join us over on Twitch! https://twitch.tv/drvem
    }
   ...
}
```

DrBot would then populate the things on it's own. I know this bot is only gonna be used for DrVem's server so it's like "does the config really matter in this case?", but I think it does, some people might also use these code bits for their own bots so its kinda nice to add some modularity to make it easier, it's also nice to keep all customizable things in the config file, imo

InimicalPart replied:
For the [next-stream-relative] we could incorporate that with /nextstream in some way, such as for example (pretty complicated but smartest), is to use the internal communication channel found in ``global.communicationChannel`` to community with the /nextstream command and retrieve information about the next stream (this does create a dependency between ``/nextstream`` and ``/mod streammessage``, which might have to be checked for (see [/src/events/wordleHandler.evt.ts](https://github.com/Incoverse/DrBot/blob/prod/src/events/wordleHandler.evt.ts#L36-L42) and [/src/commands/wordle.cmd.ts](https://github.com/Incoverse/DrBot/blob/prod/src/commands/wordle.cmd.ts#L64-L70) but you can also fetch the Twitch app token stored in ``global.twitchAppToken``)

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
In general, pretty good, and functional, but I think we could change some things.

I've added some comments along the code and added some suggested changes, here and there.

I was wondering also, maybe we could do this in a Discord embed somehow? Maybe will make things look a bit more prettier? That's up to you though.

Pretty well done for no assistance! 👏👏👏

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2024 Inimi | DrHooBs | Incoverse
```
2025 😅 
```suggestion
 * Copyright (c) 2025 Inimi | DrHooBs | Incoverse
```

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+    ): Promise<boolean> {
+        parentSlashCommand.addSubcommand((subcommand) =>
+            subcommand
+                .setName("streammessage")
```
can we add a hyphen here and everywhere else that needs it (maybe file name too)?
```suggestion
                .setName("stream-message")
```

InimicalPart replied:
This needs to be changed to work with the suggested template system. The options in "template" need to be dynamic this is hard so you can hardcode it for now and I can implement it later.

InimicalPart commented on src/commands/command-lib/mod/streammessage.cmdlib.ts:
```diff
@@ -0,0 +1,158 @@
+        });
+
+        //Get Channel List
+        const channels = await interaction.client.guilds.fetch(global.app.config.mainServer).then(guild => guild.channels.fetch())
```
Doesn't the fetch in ``guild.channels.fetch`` need await too?
```suggestion
        const channels = await (await interaction.client.guilds.fetch(global.app.config.mainServer)).channels.fetch()
```

DrHooBs committed:
Update src/commands/command-lib/mod/streammessage.cmdlib.ts

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
🎨 Improve Format

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
💄Make reply better

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
💄Fix response

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
💄 More response changes

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
Update Copyright

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
🎨 Add configurable templating

Add configurable templating for the stream message command. Also rename file and command to stream-message. This should be the final fix for this PR

InimicalPart commented:
~~Please send a snippet of the new changes to the config file, once added in the production version, merge will be approved and merged~~

Edit: Nevermind, saw the addition to config.template.jsonc. There are some changes I would want which I will add myself since they are complicated to add, after that's complete, this PR will be approved and merged.
